### DISCUSSION TITLE
Question about nightly builds

### DISCUSSION THREAD

**Discussion by vanous** [OpenTracksApp/OpenTracks#1066] [General]
I do not want to use an existing issue or pr...https://github.com/OpenTracksApp/OpenTracks/issues/952 , so let me please ask here:

> I've got the build to run on my little vserver and setup a cronjob for 1 a.m. to check for new commits and build a new nightly apk. Let's see if it is working. In case someone has already installed the first test nightly from my repo, you need to uninstall it first. I had to change the signing key.
> 
> Repo (still beta): https://fdroid.storchp.de/fdroid/repo/

@pstorch this is awesome. I am looking for inspiration on how to do a nightly release for Gadgetbridge. I do the [build already](https://codeberg.org/vanous/Gadgetbridge/releases), but am looking for an inspiration on how to do the actual repo, as a static page via CI/CD. Do you use f-droid, or repomaker periodically? Or can repomaker be used just to create the scaffolding and then simply update the xml? Thank you very much!

**Comment by pstorch** [comment]
@vanous I basically followed the instructions from here https://f-droid.org/en/docs/Setup_an_F-Droid_App_Repo/ and here https://android.izzysoft.de/articles/named/fdroid-simple-binary-repo?lang=en to install fdroid on the VM.

For each app I've created a new fdroid metadata.yml file with the nightly app id.

Then I've setup a cronjob which builds and signs the APK and copies it to the fdroid/repo dir, I also copy the fastlane metadata from the git  working dir to the fdroid metadata/app-id/ directory (excluding the title.txt file, which I've created manually). 
I also create a changelog file with the last 5 commit messages. The commit count `git rev-list HEAD --count` on the main branch is used as the version code. Ideas from @dennisguse.

Finally I run `fdroid update` and publish the fdroid/repo dir via a simple webserver.

**Reply by dennisguse** [reply]
I really love the nightly builds :)
Makes testing way easier before releases!!!

**Reply by vanous** [reply]
Thank you for the info! I will try to navigate through this.

**Reply by vanous** [reply]
This helped me a lot! 

I have a first working setup ready: https://codeberg.org/vanous/Gadgetbridge/src/branch/nightly-branch/.woodpecker.yml I still need to tweak many things...

I noticed that sometimes your repo was missing icons in the Fdroid client, do you know what happened? For me, when i run the `fdroid update` on the CI server, the icon_sha256..., featureGraphic_sha... files are erased, so i generate them locally and only update changed files via `cp`... do you notice the same?

**Reply by pstorch** [reply]
Till now I haven't realized missing icons in my repo. Someone else missing some?

**Reply by vanous** [reply]
Now i can see everything, but when i fist looked at your repo @pstorch , app icons were missing. Those files look like this: https://codeberg.org/vanous/Gadgetbridge/src/branch/pages/fdroid/repo/nodomain.freeyourgadget.gadgetbridge.nightly_nopebble/en-US

**Reply by pstorch** [reply]
I do something like this to copy the metadata files from the source repos to the f-droid repo dir:
`rsync -va --exclude 'title.txt' fastlane/metadata/android/ $FDROID_WORKDIR/metadata/$APP_ID/`

I don't think rsync is deleting the destination files first 

Maybe there was a network hickup at that moment? I see missing screenshots in the fdroid client also for the official repo quite often.

**Reply by vanous** [reply]
OK, i think i now know what the issue is. If file_sha... exists AND it was created on another filesystem (like stored/pulled via git), then fdroid will remove the file but will not recreate it anymore. I will  file an issue with fdroidserver.

**Reply by vanous** [reply]
I reported the issue [here](https://gitlab.com/fdroid/fdroidserver/-/issues/963), you would only bump into it in a setup like `pages` on Github or Codeberg, as the storage of metadata in git is what causes the issue.

@pstorch i added listing of the apps in the repo which in our case is just two, but you have several more... the code is not bulletproof (zero error handling at the moment) but the result is nice: https://vanous.codeberg.page/Gadgetbridge/fdroid/repo/ . The [generator](https://codeberg.org/vanous/Gadgetbridge/src/branch/nightly-branch/repoconfig/repo_listing.py) and the [injection](https://codeberg.org/vanous/Gadgetbridge/src/branch/nightly-branch/.woodpecker.yml#L57) into the existing index.html, which is generated by `fdroid update` (must be `pip install fdroidserver==2.1a0`).

**Reply by pstorch** [reply]
Ah, cool. I tried to test your generator, but I don't have the template index.html from fdroid. How is this generated?

**Reply by vanous** [reply]
> Ah, cool. I tried to test your generator, but I don't have the template index.html from fdroid. How is this generated?

The index.html is generated via `fdroid update` but you mast have fdroidserver 2.1a0:

`python -m pip install fdroidserver==2.1a0`

Also, make sure to install yattag via pip. It could be super easy to write this by hand though and just use f-strings... but yattag it is :)

**Reply by pstorch** [reply]
Ah, thanks. Now I got it.

**Reply by pstorch** [reply]
I guess I need to tweak it a little bit 😉
https://fdroid.storchp.de/fdroid/repo/

**Reply by vanous** [reply]
Nice! Yeah, maybe adding of markdown parsing would help :) You can change the "description" into "summary". Btw, i added changelogs again, pointing at the current list of commits.

**Reply by pstorch** [reply]
I saw the changelog link in the generator, but it gives me an error, so I commented it out:

```
Traceback (most recent call last):
  File "/home/peter/fdroid/repo_listing.py", line 69, in <module>
    href=f'{app["changelog"]}',
KeyError: 'changelog'

```

**Reply by pstorch** [reply]
Ah, one app had a missing changelog entry in the metadata. Now it works. But I think linking to the release pages is not appropriate for the nightlies. I should link to the last commits in the source repo. Even better to the last commit used to build the nightly.

**Reply by vanous** [reply]
You need to fill changelog into the config for each app, so it becomes part of the index-v1.json

 https://codeberg.org/vanous/Gadgetbridge/src/branch/nightly-branch/repoconfig/metadata/nodomain.freeyourgadget.gadgetbridge.nightly_nopebble.yml#L11

**Reply by pstorch** [reply]
Instead of the link to the releases page, I've added the content of the "whatsNew" tag (which is the changelog I created for this nightly).
Thanks again @vanous, a basic webpage was a missing piece of mine.

**Reply by vanous** [reply]
Cool idea, i will do the same 👍

**Reply by vanous** [reply]
Few updates for anyone interested.

- thank you so much for the inspiration :) it is off-topic here, but i hope this can help to some interested people
- all our setup is not fully isolated in https://codeberg.org/Freeyourgadget/fdroid-repo-config and documented as much as i could, this could help if anyone is looking into the same thing (on Codeberg but also elsewhere), so 100% of our publishing pipeline is out in the open. Tokens and keys are in CI secrets.
- i tested adding the "repo listing" directly into fdroidserver, it worked but i decided not to use it and not to push this upstream as a PR because some data were actually not available during the build. I already forgot what exactly was missing as this was few weeks back :)
- i tested using the Releases endpoints as apk sources (one must edit the index and re-sign it) but unfortunately f-droid does not and will not support this. I left the scripts in my setup repo for the record
- for the same above reasons, i stopped doing apk releases via the Releases endpoint. Also, as i now use another git repository for hosting of the pages content, the release tag was actually not relevant in that repo at all. Scripts also left in place.
